src/store/indexeddb.js

```diff
@@ -0,0 +1,403 @@
+ *     store: store,
+ * });
+ * client.startClient();
+ * client.on("sync", function(state, prevState, data) {
+ *     if (state === "PREPARED") {
+ *         console.log("Started up, now with go faster stripes!");
+ *     }
+ * });
+ * </pre>
+ *
+ * @constructor
+ * @extends MatrixInMemoryStore
+ * @param {IndexedDBStoreBackend} backend The indexed db backend instance.
+ * @param {SyncAccumulator} syncAccumulator The sync accumulator which will be
+ * loaded from IndexedDB and periodically saved to IndexedDB.
+ * @param {Object=} opts Options for MatrixInMemoryStore.
+ * @prop {IndexedDBStoreBackend} backend The backend instance. Call through to
+ * this API if you need to perform specific indexeddb actions like deleting the
+ * database.
+ */
+const IndexedDBStore = function IndexedDBStore(backend, syncAccumulator, opts) {
+    MatrixInMemoryStore.call(this, opts);
+    this.backend = backend;
+    this.startedUp = false;
+    this._syncTs = Date.now(); // updated when writes to the database are performed
+
+    // internal structs to determine deltas for syncs to the database.
+    this._userModifiedMap = {
+        // user_id : timestamp
+    };
+    this._syncAccumulator = syncAccumulator;
+
+    if (!this.backend || !this._syncAccumulator) {
+        throw new Error("Missing backend or syncAccumulator");
+    }
+};
+utils.inherits(IndexedDBStore, MatrixInMemoryStore);
+
+/**
+ * @return {Promise} Resolved when loaded from indexed db.
+  */
+IndexedDBStore.prototype.startup = function() {
+    if (this.startedUp) {
+        return q();
+    }
+    return this.backend.connect().then(() => {
+        return q.all([
+            this.backend.loadUsers(),
+            this.backend.loadAccountData(),
+            this.backend.loadSyncData(),
+        ]);
+    }).then((values) => {
+        const [users, accountData, syncData] = values;
+        console.log(
+            "Loaded data from database: sync from ", syncData,
+            " -- Reticulating splines...",
+        );
+        users.forEach((u) => {
+            this._userModifiedMap[u.userId] = u.getLastModifiedTime();
+            this.storeUser(u);
+        });
+        this.storeAccountDataEvents(accountData);
+        this._syncTs = Date.now(); // pretend we've written so we don't rewrite
+        this.setSyncToken(syncData.nextBatch);
+        this._setSyncData(syncData.nextBatch, syncData.roomsData);
+    });
+};
+
+/**
+ * Return the accumulator which will have the initial /sync data when startup()
+ * is called.
+ * @return {SyncAccumulator}
+ */
+IndexedDBStore.prototype.getSyncAccumulator = function() {
+    return this._syncAccumulator;
+};
+
+/**
+ * Possibly write data to the database.
+ * @return {Promise} Promise resolves after the write completes.
+ */
+IndexedDBStore.prototype.save = function() {
+    const now = Date.now();
+    if (now - this._syncTs > WRITE_DELAY_MS) {
+        return this._syncToDatabase().catch((err) => {console.error("sync fail:", err);});
+    }
```

Comment:
In the worst case if the timer doesn't fire, we will update when the HTTP response comes back (since that's a held-open connection rather than using `setTimeout` so isn't affected by the throttling).